go/enclave/services/sequencer.go

```diff
@@ -0,0 +1,329 @@
+	mempool           mempool.Manager
+	storage           db.Storage
+	encryption        crypto.TransactionBlobCrypto
+
+	// This is used to coordinate creating
+	// new batches and creating fork batches.
+	batchProductionMutex sync.Mutex
+}
+
+func NewSequencer(
+	consumer components.L1BlockProcessor,
+	producer components.BatchProducer,
+	registry components.BatchRegistry,
+	rollupProducer components.RollupProducer,
+	rollupConsumer components.RollupConsumer,
+
+	logger gethlog.Logger,
+
+	hostID gethcommon.Address,
+	chainConfig *params.ChainConfig,
+	enclavePrivateKey *ecdsa.PrivateKey, // this is a key known only to the current enclave, and the public key was shared with everyone during attestation
+	mempool mempool.Manager,
+	storage db.Storage,
+	encryption crypto.TransactionBlobCrypto,
+) Sequencer {
+	return &sequencer{
+		blockConsumer:     consumer,
+		batchProducer:     producer,
+		batchRegistry:     registry,
+		rollupProducer:    rollupProducer,
+		rollupConsumer:    rollupConsumer,
+		logger:            logger,
+		hostID:            hostID,
+		chainConfig:       chainConfig,
+		enclavePrivateKey: enclavePrivateKey,
+		mempool:           mempool,
+		storage:           storage,
+		encryption:        encryption,
+	}
+}
+
+func (s *sequencer) CreateBatch() (*core.Batch, error) {
+	s.batchProductionMutex.Lock()
+	defer s.batchProductionMutex.Unlock()
+
+	hasGenesis, err := s.batchRegistry.HasGenesisBatch()
+	if err != nil {
+		return nil, fmt.Errorf("unknown genesis batch state. Cause: %w", err)
+	}
+
+	// L1 Head is only updated when isLatest: true
+	// when a block is specified it will override this and allow
+	// building batches for unfinished forks.
+	block, err := s.blockConsumer.GetHead()
+	if err != nil {
+		return nil, fmt.Errorf("failed retrieving l1 head. Cause: %w", err)
+	}
+
+	if !hasGenesis {
+		return s.initGenesis(block)
+	}
+
+	return s.createNewHeadBatch(block)
+}
+
+// TODO - This is iffy, the producer commits the stateDB
+func (s *sequencer) initGenesis(block *common.L1Block) (*core.Batch, error) {
+	batch, msgBusTx, err := s.batchProducer.CreateGenesisState(block.Hash(), s.hostID, uint64(time.Now().Unix()))
+	if err != nil {
+		return nil, err
+	}
+
+	if err := s.mempool.AddMempoolTx(msgBusTx); err != nil {
+		return nil, fmt.Errorf("failed to queue message bus creation transaction to genesis. Cause: %w", err)
+	}
+
+	if err := s.signBatch(batch); err != nil {
+		return nil, fmt.Errorf("failed signing created batch. Cause: %w", err)
+	}
+
+	if err := s.batchRegistry.StoreBatch(batch, nil); err != nil {
+		return nil, fmt.Errorf("failed storing batch. Cause: %w", err)
+	}
+
+	return batch, nil
+}
+
+func (s *sequencer) createNewHeadBatch(block *common.L1Block) (*core.Batch, error) {
+	headBatch, err := s.batchRegistry.GetHeadBatch()
+	if err != nil {
+		return nil, err
+	}
+
+	// We find the ancestral batch for the block as it might be different
```

Comment:
I have a feeling this logic is unnecessarily complex.
I'll try to restate the problem.

Problem: I want to add a batch to the l2 chain. 

Logic:
1. If the current l2 head points to an ancestor of the current l1 head, then I just carry on.

2. if the current l2 head points to a fork of the l1 head, then I go to the l1 block where the l1 fork happened. 
Try to find batches in the db for the current canonical fork ( maybe there are some already). If there are, then carry on from where they left off, but making sure to duplicate batches up to the maximum height already created.


Based on the above description. I'd say that a very important parameter of the sequencer is the "MaxL2Height". Basically if you're creating a batch with a height of less than that, you have to make sure to duplicate the already existing batch at that height.